File: c++/cote/GJM/시계_맞추기.cpp
// 에러 미해결

#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#define INF 9999

using namespace std;

vector<int>& getClocks(const string& s) {
  vector<int>* result = new vector<int>();
  stringstream ss(s);
  string sbuf;

  for (int i = 0; i < 12; i++) {
    getline(ss, sbuf, ' ');
    result->push_back(stoi(sbuf));
  }
  return *result;
}

bool aligned(vector<int>& clocks) {
  for (int i = 0; i < 16; i++) {
    if (clocks[i] != 12) return false;
  }

  return true;
}

void push(const char linked[10][16 + 1], vector<int>& clocks, int cur) {
  for (int i = 0; i < 16; i++) {
    if (linked[cur][i] == 'x') {
      clocks[i] += 3;
      if (clocks[i] == 15) clocks[i] = 3;
    }
  }
}

int solve(const char linked[10][16 + 1], vector<int>& clocks, int cur) {
  if (cur == 10) {
    return aligned(clocks) ? 0 : INF;
  }

  int ret = INF;

  for (int i = 0; i < 4; i++) {
    ret = min(ret, i + solve(linked, clocks, cur + 1));
    push(linked, clocks, cur);
  }

  return ret;
}

int main(void) {
  string buf;
  fstream fd("data/시계_맞추기.txt");

  const char linked[10][16 + 1] = {"xxx.............", "...x...x.x.x....",
                                   "....x.....x...xx", "x...xxxx........",
                                   "......xxx.x.x...", "x.x...........xx",
                                   "...x..........xx", "....xx.x......xx",
                                   ".xxxxx..........", "...xxx...x...x.."};

  getline(fd, buf);
  int test_cases = stoi(buf);

  while (test_cases > 0) {
    getline(fd, buf);
    vector<int> clocks = getClocks(buf);

    cout << solve(linked, clocks, 0) << endl;

    test_cases--;
  }

  return 0;
}

File: go/cote/GJM/시계_맞추기.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const INF = 987654321

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func aligned(clocks []int) bool {

	for _, clock := range clocks {
		if clock != 12 {
			return false
		}
	}

	return true
}

func push(cur int, linked [][]bool, clocks []int) {
	for i, v := range clocks {
		if linked[cur][i] == true {
			clocks[i] = (v + 3) % 12
		}
	}
}

func search(cur int, linked [][]bool, clocks []int) int {
	if cur == len(linked) {
		if aligned(clocks) {
			return 0
		}
		return INF
	}

	ret := INF

	for it := 0; it < 4; it++ {
		ret = min(ret, it+search(cur+1, linked, clocks))
		push(cur, linked, clocks)
	}

	return ret
}

func solve(linked [][]bool, clocks []int) int {
	return search(0, linked, clocks)
}

func main() {
	f, _ := os.Open("data/시계_맞추기.txt")
	r := bufio.NewReader(f)
	s, _ := r.ReadString('\n')

	l := strings.Split(strings.Trim(s, "\n"), " ")

	numButtons, _ := strconv.Atoi(l[0])
	numClocks, _ := strconv.Atoi(l[1])
	linked := make([][]bool, numButtons)

	for i, _ := range linked {
		s, _ = r.ReadString('\n')
		linked[i] = make([]bool, numClocks)
		s = strings.Trim(s, "\n")

		for j, v := range s {
			switch v {
			case 'x':
				linked[i][j] = true
			default:
				linked[i][j] = false
			}
		}
	}

	s, _ = r.ReadString('\n')
	testcases, _ := strconv.Atoi(strings.Trim(s, "\n"))
	for it := 0; it < testcases; it++ {
		s, _ = r.ReadString('\n')
		s = strings.Trim(s, "\n")

		clocks := make([]int, numClocks)
		states := strings.Split(s, " ")

		for j, v := range states {
			time, _ := strconv.Atoi(v)
			clocks[j] = time
		}
		fmt.Println(solve(linked, clocks))
	}
}
